crates/extensions/tedge_config_manager/src/config.rs

```diff
@@ -0,0 +1,253 @@
+use log::error;
+use log::info;
+use log::warn;
+use serde::Deserialize;
+use std::borrow::Borrow;
+use std::collections::HashSet;
+use std::fs;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
+use tedge_config::ReadError;
+use tedge_mqtt_ext::TopicFilter;
+use tedge_utils::file::PermissionEntry;
+
+use super::error::InvalidConfigTypeError;
+
+pub const DEFAULT_PLUGIN_CONFIG_FILE_NAME: &str = "tedge-configuration-plugin.toml";
+pub const DEFAULT_OPERATION_DIR_NAME: &str = "plugins/";
+pub const DEFAULT_PLUGIN_CONFIG_TYPE: &str = "tedge-configuration-plugin";
+
+/// Configuration of the Configuration Manager
+#[derive(Clone, Debug)]
+pub struct ConfigManagerConfig {
+    pub config_dir: PathBuf,
+    pub mqtt_topic_root: String,
+    pub mqtt_device_topic_id: String,
+    pub plugin_config_dir: PathBuf,
+    pub plugin_config_path: PathBuf,
+    pub config_reload_topics: TopicFilter,
+    pub config_update_topic: TopicFilter,
+    pub config_snapshot_topic: TopicFilter,
+}
+
+pub struct ConfigManagerOptions {
+    pub config_dir: PathBuf,
+    pub mqtt_topic_root: Arc<str>,
+    pub mqtt_device_topic_id: Arc<str>,
+}
+
+impl ConfigManagerConfig {
+    pub fn from_options(cliopts: ConfigManagerOptions) -> Result<Self, ReadError> {
+        let config_dir = cliopts.config_dir;
+        let mqtt_topic_root = cliopts.mqtt_topic_root;
+        let mqtt_device_topic_id = cliopts.mqtt_device_topic_id;
+
+        let plugin_config_dir = config_dir.join(DEFAULT_OPERATION_DIR_NAME);
+        let plugin_config_path = plugin_config_dir.join(DEFAULT_PLUGIN_CONFIG_FILE_NAME);
+
+        let config_reload_topics: TopicFilter = vec![
+            format!(
+                "{}/{}/cmd/config_snapshot",
+                mqtt_topic_root, mqtt_device_topic_id
+            ),
+            format!(
+                "{}/{}/cmd/config_update",
+                mqtt_topic_root, mqtt_device_topic_id
+            ),
+        ]
+        .try_into()
+        .expect("Infallible");
```

Comment:
Actually, "infallible" is wrong. If user gives such command line arguments as `--mqtt-device-topic-id foo/bar`, `--mqtt-device-topid-id foo+bar`, they break the topic scheme.

This means, validation for the command line input for `mqtt_topic_root` and `mqtt_device_topic_id` is required (most likely inside of `main.rs`). Also, I believe that we have similar validation implementation in the EntityStore.